pkg/controller/inferenceservice/reconcilers/knative/service_reconciler.go

```diff
@@ -84,51 +67,47 @@ func (r *ServiceReconciler) reconcileCanary(isvc *v1alpha2.InferenceService) err
 }
 
 func (r *ServiceReconciler) reconcileComponent(isvc *v1alpha2.InferenceService, component constants.InferenceServiceComponent, isCanary bool) error {
+	endpointSpec := &isvc.Spec.Default
+	serviceName := constants.DefaultServiceName(isvc.Name, component)
+	propagateStatus := isvc.Status.PropagateDefaultStatus
 	if isCanary {
-		if isvc.Spec.Canary == nil {
-			if err := r.finalizeCanaryService(isvc, component); err != nil {
-				return err
-			}
-			isvc.Status.PropagateCanaryStatus(component, nil)
-			return nil
+		endpointSpec = isvc.Spec.Canary
+		serviceName = constants.CanaryServiceName(isvc.Name, component)
+		propagateStatus = isvc.Status.PropagateCanaryStatus
+	}
+	var service *knservingv1alpha1.Service = nil
```

Comment:
Why nil this out?

Reply:
yes, we can just  change it to`var service *knservingv1alpha1.Service` since pointer default value is `nil`
done

Reply:
looks like the check `if endpointSpec != nil` is unnecessary here, basically it can not be nil. Removing this check will make code a lot cleaner.

Reply:
`isvc.Spec.Canary` can be nil. when isCanary,  endpointSpec = isvc.Spec.Canary

Reply:
Can you make the check ` isvc.Spec.Canary != nil `before calling `reconcileComponent(isvc, component, true)`

Reply:
why check `isvc.Spec.Canary != nil` before calling `reconcileComponent(isvc, component, true)`? what if `isvc.Spec.Canary != nil` or `isvc.Spec.Canary == nil`

reconcileComponent can handle two cases: create knatvie service or delete knatvie service

Reply:
Good point :)